### DISCUSSION TITLE
RuntimeError: Sizes of tensors must match except in dimension 1. Got 47 and 48 in dimension 4 (The offending index is 1)

### DISCUSSION THREAD

**Discussion by talhaanwarch** [Project-MONAI/MONAI#2319] [Q&A]
Here is the model
from monai.networks.nets import UNet
from monai.networks.layers import Norm
```
model = UNet(
            dimensions=3,
            in_channels=1,
            out_channels=5,
            channels=(16, 32, 64, 128, 256),
            strides=(2, 2, 2, 2),
            num_res_units=2,
            norm=Norm.BATCH)
```
When I pass the data
```
import torch
x=torch.randn(1, 1, 256, 256,94)
out=model(x)
out.shape
```
I got this error
`RuntimeError: Sizes of tensors must match except in dimension 1. Got 47 and 48 in dimension 4 (The offending index is 1)`

I don't think padding 94 to 96 is a solution. because I want output shape to be `(1, 1, 256, 256,94)`

**Comment by Nic-Ma** [comment]
Hi @talhaanwarch ,

Thanks for your interest here.
Although you don't want to pad 94 to 96, it is the root cause.
Because here you set `channels=(16, 32, 64, 128, 256)` in `UNet`, so the input size will be divided **4** times by **2** in the net, so at least the size should be **N x 16**.
@ericspod @wyli  please feel free to correct me if I misunderstand the root cause.

Thanks.

**Reply by talhaanwarch** [reply]
In #2218 I was recommended to use batch size of 1 to encounter image of different slice length (z-axis). So after this error, I think it's not feasible

**Reply by rijobro** [reply]
Changing the batch size helps when you have images of different sizes. In your example code, there is only one image so it's not related to batch size. As @Nic-Ma says, your input image needs to be divisible by 2^4 in all dimensions.

If you want your output to be reverted to the initial size (e.g., 94 in z-direction), you can use `DivisiblePadd` (94->96) prior to passing your image through the network and then use the inverse method after passing through the model to revert to back to the original size (96->94).

**Reply by Nic-Ma** [reply]
@rijobro , Using `Invertd` transforms is a good idea!
Thanks.

**Reply by talhaanwarch** [reply]
thanks @rijobro and @Nic-Ma  . But in the segmentation task, let say my image  (94 in z-direction) after `DivisiblePadd` (96 in z-direction) is passed to a network, it produces a mask of  (96 in z-direction) and my actual mask is 94 in z-direction. So i am supposed to invert my predicted mask to 94 and then compute the loss. 
I got this from your above message. Am i correct?

**Reply by talhaanwarch** [reply]
i want to do the same as in the spline tutorial, but I want to keep the predicted mask (segment) shape as the input shape, let say `(512,512,58)`
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d_lightning.ipynb

**Reply by rijobro** [reply]
You shouldn't need to invert to be able to calculate the loss. During training you can pad both the input image and the ground truth segmentation. Since they both get padded by the same amount, the loss function can be calculated directly using the padded ground truth segmentation and the output of your network. If you decide you want to save your results to file, for example, then you can think about doing the inverse at that point.

**Reply by rijobro** [reply]
I haven't spent much time looking at it, but I imagine the notebook you just linked would work just fine if we added a `DivisiblePad` to the list of transforms.

**Reply by talhaanwarch** [reply]
thanks @rijobro  i am confused what should be the value of K 
`DivisiblePadd(keys=["image", "label"],k=...),`

**Reply by rijobro** [reply]
The stride of your convolutions will determine the step of downsampling at each layer. If your network definition has `strides=(2, 2, 2, 2)`, then your images need to be divisible by 2^4=16. Hope that helps.
